WalletWasabi/EventSourcing/InMemoryEventRepository.cs

```diff
@@ -0,0 +1,216 @@
+	/// </summary>
+	public class InMemoryEventRepository : IEventRepository
+	{
+		protected static readonly IReadOnlyList<WrappedEvent> EmptyResult = ImmutableList<WrappedEvent>.Empty;
+		protected static readonly IReadOnlyList<string> EmptyIds = ImmutableList<string>.Empty;
+
+		private readonly ConcurrentDictionary
+			// aggregateType
+			<string,
+			ConcurrentDictionary
+				// aggregateId
+				<string,
+				(
+					// SequenceId of the last event of this aggregate
+					long TailSequenceId,
+
+					// Ordered list of events
+					ImmutableList<WrappedEvent> Events
+				)>> _aggregatesEventsBatches = new();
+
+		private readonly ConcurrentDictionary
+			// aggregateType
+			<string,
+			(
+				// Index of the last aggregateId in this aggregateType
+				long TailIndex,
+
+				// List of aggregate Ids in this aggregateType
+				ImmutableSortedSet<string> Ids
+			)> _aggregatesIds = new();
+
+		public Task AppendEventsAsync(
+			string aggregateType,
+			string aggregateId,
+			IEnumerable<WrappedEvent> wrappedEvents)
+		{
+			Guard.NotNullOrEmpty(nameof(aggregateType), aggregateType);
+			Guard.NotNullOrEmpty(nameof(aggregateId), aggregateId);
+			Guard.NotNull(nameof(wrappedEvents), wrappedEvents);
+			var wrappedEventsList = wrappedEvents.ToList().AsReadOnly();
+			if (wrappedEventsList.Count <= 0)
+			{
+				return Task.CompletedTask;
+			}
+			var firstSequenceId = wrappedEventsList[0].SequenceId;
+			var lastSequenceId = wrappedEventsList[^1].SequenceId;
+			if (firstSequenceId <= 0)
+			{
+				throw new ArgumentException("First event sequenceId is not natural number.", nameof(wrappedEvents));
+			}
+			if (lastSequenceId <= 0)
+			{
+				throw new ArgumentException("Last event sequenceId is not natural number.", nameof(wrappedEvents));
+			}
+			if (lastSequenceId - firstSequenceId + 1 != wrappedEventsList.Count)
+			{
+				throw new ArgumentException("Event sequence ids are out of whack.", nameof(wrappedEvents));
+			}
+
+			var aggregateEventsBatches = _aggregatesEventsBatches.GetOrAdd(aggregateType, _ => new());
+			var (tailSequenceId, events) = aggregateEventsBatches.GetOrAdd(
+				aggregateId,
+				_ => (0, ImmutableList<WrappedEvent>.Empty));
+
+			if (tailSequenceId + 1 < firstSequenceId)
+			{
+				throw new ArgumentException($"Invalid firstSequenceId (gap in sequence ids) expected: '{tailSequenceId + 1}' given: '{firstSequenceId}'.", nameof(wrappedEvents));
+			}
+
+			// no action
+			Validated();
+
+			var newEvents = events.AddRange(wrappedEventsList);
+
+			// Atomically detect conflict and replace lastSequenceId and lock to ensure strong order in eventsBatches.
+			if (!aggregateEventsBatches.TryUpdate(
+				key: aggregateId,
+				newValue: (lastSequenceId, newEvents),
+				comparisonValue: (firstSequenceId - 1, events)))
+			{
+				Conflicted(); // no action
+				throw new OptimisticConcurrencyException($"Conflict while commiting events. Retry command. aggregate: '{aggregateType}' id: '{aggregateId}'");
```

Comment:
It seems after this exception is thrown the caller should retry and in that case we don't know how many times it could retry in the worst case (there will be a limit ofc) but it could be good to avoid throwing exceptions.

Reply:
retry if any is responsibility of a caller, it is irrelevant on his level.

responsibility of repository is to throw in case of failure